### PR TITLE
miri: Memory data revived, Hooks for stack frame push/pop

### PR DESCRIPTION
**Pull request by RalfJung**
r? @oli-obk

### CODE REVIEW

File: src/librustc/mir/interpret/allocation.rs [hunk 1]
```diff
@@ -49,13 +49,19 @@ pub struct Allocation<Tag=(),Extra=()> {
     pub extra: Extra,
 }
 
-pub trait AllocationExtra<Tag>: ::std::fmt::Debug + Default + Clone {
+pub trait AllocationExtra<Tag, MemoryExtra>: ::std::fmt::Debug + Clone {
+    /// Hook to initialize the extra data when an allocation gets crated.
```
Comment (oli-obk) [inline src/librustc/mir/interpret/allocation.rs]:
```suggestion
    /// Hook to initialize the extra data when an allocation gets created.
```

**Review by oli-obk — APPROVED** [review]

**Comment by RalfJung** [comment]
Rebased.

@bors r=oli-obk

**Comment by RalfJung** [comment]
Rebased.

@bors r=oli-obk

**Comment by RalfJung** [comment]
Rebased (4th time).

@bors r=oli-obk

**Comment by kennytm** [comment]
```
[01:27:33] error[E0107]: wrong number of type arguments: expected 2, found 1
[01:27:33]    --> src/tools/miri/src/stacked_borrows.rs:362:6
[01:27:33]     |
[01:27:33] 362 | impl AllocationExtra<Borrow> for Stacks {
[01:27:33]     |      ^^^^^^^^^^^^^^^^^^^^^^^ expected 2 type arguments
[01:27:33] 
[01:27:33] error: aborting due to previous error
```

**Pull request merged**